### PR TITLE
feat(storage-browser): add initial support for locationActions

### PR DESCRIPTION
**Pull request by calebpollman**
#### Description of changes

Add initial `actions` support

> allows for providing of `actions` but still `LocationActionView` continues to use underlying `actions`. Complete support in follow-ups

- refactor `ActionsSelect` component to `ActionsMenu` (generic component) and `ActionsMenuControl`
- simplify `ACTIONS_SELECT` state in preparation of migration to `LocationActions` state
- update `UploadControls` to contain selected `action` lookup logic in preparation of migration to moving all logic to parent `LocationActionView`
- refactor file selection UX 
  - migrate file/folder selection to `LocationActionView` 
  - localize `File` selection state to `LocationActionView`

#### Issue #, if available
NA

#### Description of how you validated changes
Update/add unit tests, manual tests
#### Checklist

- [x] Have read the [Pull Request Guidelines](https://github.com/aws-amplify/amplify-ui/blob/main/CONTRIBUTING.md)
- [x] PR description included
- [x] `yarn test` passes and tests are updated/added
- [x] PR title and commit messages follow [conventional commit](https://www.conventionalcommits.org/en/v1.0.0/#summary) syntax

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/react-storage/src/components/StorageBrowser/context/controls/locationActions/defaults.tsx [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+};
+
+const UPLOAD_FILES: LocationAction = {
+  handler,
+  options: {
+    disable: (items) => !!items.length,
+    displayName: 'Upload File',
+    hide: (permission) => permission === 'READ',
+    icon: <IconElement variant="upload-file" />,
+    selectionData: ['file', '.png'],
```
Comment (timngyn) [inline packages/react-storage/src/components/StorageBrowser/context/controls/locationActions/defaults.tsx:38]: What's the reason for having `.png` as the default selection data?

File: packages/react-storage/src/components/StorageBrowser/Views/LocationActionView/UploadControls.tsx [hunk 2]
```diff
@@ -136,18 +140,45 @@ const renderRowItem: RenderRowItem<CancelableTask> = (row, index) => {
...
+  const accept = type && Array.isArray(value) ? value[1] : undefined;
+
+  return { type, accept };
+};
+
 export const UploadControls = (): JSX.Element => {
-  const [state, handleUpdateState] = useControl({
+  const [{ history, path }] = useControl({ type: 'NAVIGATE' });
+  const [files, setFiles] = React.useState<File[]>([]);
+  const [fileSelect, handleSelect] = useFileSelect(setFiles);
```
Comment (timngyn) [inline packages/react-storage/src/components/StorageBrowser/Views/LocationActionView/UploadControls.tsx:161]: I'm not sure if this is capturing the `webkitRelativePath` property that includes the folder names. As a result I think the upload folder action is uploading only individual files and not saving the folders. Should this be happening?

File: packages/react-storage/src/components/StorageBrowser/Views/LocationActionView/UploadControls.tsx [hunk 3]
```diff
@@ -216,13 +247,13 @@ export const UploadControls = (): JSX.Element => {
...
+      {fileSelect}
       <Title />
-      <Exit onClick={() => handleUpdateState({ type: 'EXIT' })} />
+      <Exit onClick={() => handleUpdateState({ type: 'CLEAR' })} />
       <Primary
         onClick={() => {
-          if (!items) return;
           handleUpload();
         }}
       >
```
Comment (timngyn) [inline packages/react-storage/src/components/StorageBrowser/Views/LocationActionView/UploadControls.tsx:259]: not blocking: Should the primary button be disabled after we start the upload action? At the moment if I have selected files, I can keep clicking the primary button and the network request gets sent each time even though the items already finished uploading

File: packages/react-storage/src/components/StorageBrowser/context/controls/locationActions/defaults.tsx (hunk 1 above)
Comment (calebpollman) [reply packages/react-storage/src/components/StorageBrowser/context/controls/locationActions/defaults.tsx:38]: Just a typo from testing :)

File: packages/react-storage/src/components/StorageBrowser/components/ActionsMenu.tsx [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+      <ViewElement
+        aria-label="Actions"
+        className={`${MENU_CLASS_NAME}${
+          isOpen ? ` ${MENU_CLASS_NAME}--open` : ''
+        }`}
+        data-testid="ACTIONS_MENU_LIST"
+        role="menu"
+        variant="actions-menu-list"
+      >
+        {data?.map(renderItem)}
```
[the remark this answers is not in the document]
Comment (calebpollman) [reply packages/react-storage/src/components/StorageBrowser/components/ActionsMenu.tsx:72]: It's due to the `children` passed to `renderItem` being in an array without keys, updating

File: packages/react-storage/src/components/StorageBrowser/Views/LocationActionView/UploadControls.tsx (hunk 2 above)
Comment (calebpollman) [reply packages/react-storage/src/components/StorageBrowser/Views/LocationActionView/UploadControls.tsx:161]: That was not in scope of the initial approach for `useFileSelect` but can be added in a seperate PR

File: packages/react-storage/src/components/StorageBrowser/Views/LocationActionView/UploadControls.tsx (hunk 3 above)
Comment (calebpollman) [reply packages/react-storage/src/components/StorageBrowser/Views/LocationActionView/UploadControls.tsx:259]: Oh it definitely should be, will update

File: packages/react-storage/src/components/StorageBrowser/components/ActionsMenu.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (calebpollman) [reply packages/react-storage/src/components/StorageBrowser/components/ActionsMenu.tsx:72]: Addressed here: https://github.com/aws-amplify/amplify-ui/pull/5671/files#diff-9fcb5313af60f68e9d679c42ab3354ef11619211273d15843df478dab9a5541bR33

**Review by reesscot — APPROVED** [review]

**Review by timngyn — APPROVED** [review]

**Pull request merged**